### ISSUE TITLE
Potential ways to improve metrics

### ISSUE THREAD

**Issue by brian-brazil** [mwitkow/go-grpc-prometheus#2]
There's a few ways these metrics can be improved in line with [best practices]( https://prometheus.io/docs/instrumenting/writing_exporters/)

`service` is a common target label, so it is unwise to use it as an instrumentation label as it'll cause clashes for some users. I recommend `grpc_service` instead.

`type` is a very generic label name, try to find something more specific. This is to avoid half the metrics exposed by a binary having a `type` label, each with different meanings.

`grpc_server_rpc_handled_bucket` should include the seconds unit in the name

`grpc_server_rpc_handled_bucket` should not include `code` as a label. There's two reasons. Firstly "The list of all statuses is to long", which combined with the other 4 labels in this metric threatens to make the metric cardinality too high. Secondly is it strongly recommend not to break out latency metrics by success/failure. Better would be a separate failures metric broken out by code, don't forget to initialise all possible label values to 0.

**Comment by mwitkow** [comment]
Re `type`, yea this may need a better name. There's no clear indication on the user side what this "is". I went with type because I couldn't think about anything else that would distinguish `unary`, `client_stream`, `server_stream` and `bidi_stream`. 
`class`, `convention` maybe? ` @ejona86 thoughts?

Re `service`: I'm kinda strongly for this. `service MyService` is what users type in `.proto` files, and seems to be a first-level concept in upstream documentation (see [service defintion](http://www.grpc.io/docs/guides/concepts.html#service-definition)).

Re: `grpc_server_rpc_handled_bucket`.. for seconds, is it `_sec` or `_seconds`? 

As to not including `code` as a label. There's a tradeoff between functionality and cardinality. The cardinality multiplier is only 14x (but more like 3-4x in practice, given our use case, since rarely all codes are used). Over a year in production, we did find having a code-breakdown of latency metrics useful. For example we found caching issues based on quantiles of `code.NotFound`.

**Comment by brian-brazil** [comment]
> seems to be a first-level concept in upstream documentation. Prefixing it with grpc_ would stand out weirdly among all other labels, and the meaning is obvious since the metric deals with grpc itself.

The problem is it's a first-level concept in many users' architecture too, so it'll end up as `exported_service` for a subset of your users which is worse. You could prefix method too.

> is it _sec or _seconds

We usually use seconds.

> The cardinality multiplier is only 14x (but more like 3-4x in practice, given our use case, since rarely all codes are used). 

It's 14x on top of a histogram, and every method+service the user has. It's the combinatorial explosion that's the risk. That it works for you doesn't mean it'll work for all your users.

> (but more like 3-4x in practice, given our use case, since rarely all codes are used). 

This is also a problem, as it'll break common expressions if there hasn't been an error or success yet. All 14 need to be always exported.

More concretely, this breaks your "unary request error rate" and "unary request error percentage" examples.

> Over a year in production, we did find having a code-breakdown of latency metrics useful. For example we found caching issues based on quantiles of code.NotFound.

I'm not saying it's not useful in specific rare circumstances, however on balance it's much more likely to cause problems than help in both terms of functionality and cardinality.

For example your "average response stream size" example is incorrect as the denominator doesn't include all requests. The more common failure mode is that users will only graph successful latency, and due to that miss that their increased latency is due to slow failures.

Overall this is an anti-pattern. At some point we have to accept that there's a limit to how much data you can send out via metrics, and switch to logs instead. For a general library like this the worst case must be considered over the best case where the data happened to be useful for you, and wasn't too big to put strain on your monitoring, and you were smart enough to avoid the various pitfalls of using this type of breakout.

Putting a 14X multiplier on your already biggest metric in a way that has known usability issues is not something to be done lightly. It might be acceptable as something you can opt-in to as an additional metric, but shouldn't be a default.

"unary request error percentage" is also incorrect, as it doesn't consider in-flight requests.

**Comment by fabxc** [comment]
Feeling both sides on `service`. But I do believe most people should assign this target label to `job` instead. Often people have their `job` be on opaque string of the monitored service and additional dimensions, which I don't consider good.
Practically, using `grpc_service` may prevent some users from the `exported_service` weirdness but this way also encourages non-optimal configuration.

Can we make the `code` optional? We have it for HTTP requests as well and it turned out useful. But the blow up might be disproportionate for some setups. So I think it should be the user's call.

**Comment by brian-brazil** [comment]
> But I do believe most people should assign this target label to job instead.

That depends on the use case. Once you get beyond the trivial, you need more labels than `job` to organise your targets and Prometheus servers. `service` and `team` are likely to be the first labels used for this.

> But the blow up might be disproportionate for some setups.

If we consider the scale of setups I've seen in the past we'd be talking 1k instances with 50 services which is just over 10M time series. That's too much to fit inside a single Prometheus from just a single metric, and 1k instances should fit inside a single Prometheus nicely.

Even with a single service that's 210 time series, which is definitely on the high side considering the recommendation is to keep a full cardinality metric below 10 and definitely below 100.

> So I think it should be the user's call.

It'd need to be a separate metric, otherwise expressions can't be shared.

**Comment by fabxc** [comment]
> That depends on the use case. Once you get beyond the trivial, you need more labels than job to organise your targets and Prometheus servers. service and team are likely to be the first labels used for this.

Yes, exactly my point. Not going with `service` is certainly the safer bet.
Given that `job` mostly partitions metrics with equal semantics/avoids conflicts it's equivalent to `service` though in most sane cases I can think of though.

**Comment by mwitkow** [comment]
I'd rather not use `grpc_` prefixes for labels of metrics that already are prefixed with `grpc`. It feels like clutter. I also think that most users of gRPC will be interested in "services" on gRPC level, i.e. it doesn't really matter in which `job` your code is running in, what is important is what `service` it implements. In other words, I believe that the gRPC concept of service is going to be of higher importance to users than scrape targets.

> For example your "average response stream size" example is incorrect as the denominator doesn't include all requests. The more common failure mode is that users will only graph successful latency, and due to that miss that their increased latency is due to slow failures.

The denominator counts all completed responses (successful and failred), while the nominator counts messages in all responses total. The reason why it's counting successful and failed responses is due to the nature of gRPC streams: you can return 5 messages in your response and then encounter a `code.Internal` you'd return tot he caller. Can you elaborate how is that incorrect?

>Better would be a separate failures metric broken out by code, don't forget to initialise all possible label values to 0.

This is not possible, unfortunately. There is no way you can iterate services registered with `grpc.Server` and you can initialized them to 0. As such it can only be done on request.

> This is also a problem, as it'll break common expressions if there hasn't been an error or success yet. All 14 need to be always exported.
> More concretely, this breaks your "unary request error rate" and "unary request error percentage" examples.

I'm not arguing they shouldn't, I'm arguing that since we can't initialize everything to 0, only certain often used codes will be introduced. Can you elaborate how they "break" if they're not exported? 

> Can we make the code optional? We have it for HTTP requests as well and it turned out useful. But the blow up might be disproportionate for some setups. So I think it should be the user's call.

Can you point me as to how you made it optional? Alternatively, at Improbable we have a grouping of response codes by "cause": whether this is a user error, system error or transport error. This is still useful for SLA tracking but has lower cardinality.

To summarize so far:
 * I'm for keeping `service`, since I feel gRPC users would care about it more than about scrape targets.
 * I'll add `_seconds` suffix to the handling bucket.
 * `type` seems wonky, but I still don't have a better candidate
 * one cannot export `0` on initialization due to technical reason
 * still to be convinced that `codes` are bad, the cardinality isn't that huge even for our mediocre use case since the `service`, `method` and `code` are static index labels
 * consider adding `cause` instead of `code` and make `code` optional

**Comment by brian-brazil** [comment]
> I believe that the gRPC concept of service is going to be of higher importance to users than scrape targets.

That only works if the user starts using gRPC metrics before they start using the `service` label for other purposes, that's not at all a safe assumption.

Remember that you are only one of potentially hundreds of libraries exposing Prometheus metrics within an organisation. It's not reasonable for users to know in advance all the libraries they might use, and what common target labels they've staked a claim on.

> The denominator counts all completed responses (successful and failed), while the nominator counts messages in all responses total.

The denominator is only including successes.

> This is not possible, unfortunately. There is no way you can iterate services registered with grpc.Server and you can initialized them to 0. As such it can only be done on request.

That's a bit of a problem then. You could initialise on each rpc, though that'd have a performance impact.

> Can you elaborate how they "break" the example? Is it because missing labels?

It's time series that are sometimes missing.  If you have had only successes you'll get nothing rather than 0.

> still to be convinced that codes are bad, the cardinality isn't that huge even for our mediocre use case since the service, method and code are static index labels

The cardinality is massive here. Even offering a histogram without code is questionable.

> Can you point me as to how you made it optional?

It'd be a separate parameter into your library to turn the metric on/off.

> consider adding cause instead of code and make code optional

Cause is still a bit big, and there's a lot of vagueness here (are not found and resource exhaustion user or system errors?)

**Comment by mwitkow** [comment]
> > consider adding cause instead of code and make code optional
> Cause is still a bit big, and there's a lot of vagueness here (are not found and resource exhaustion user or system errors?)

Back at Google we did build SLA models based on gRPC return codes. The definition of codes is fairly clear now (e.g. not found is a client error, similar to a `404`). 

> The cardinality is massive here. Even offering a histogram without code is questionable.

I really do not agree with your assessment. Neither of the other parameters is dynamic, and the labels are stable (service names and methods). I agree there's a lot of data to crunch, but RPC monitoring deals with 80% of user-provided monitoring needs. We both know from our previous SRE experience how important good RPC metrics are, as developers often forget to instrument their code in other fashion.


> > Can you point me as to how you made it optional?
> It'd be a separate parameter into your library to turn the metric on/off.

And not initialize the variables in `func init` of the package? Can can you give an example of how you guys do it?

> The denominator counts all completed responses (successful and failed), while the nominator counts messages in all responses total.
> The denominator is only including successes.

Ah, good point. I thought I didn't have `code="OK"`. Fixed. Thanks for pointing it out!

**Comment by brian-brazil** [comment]
> The definition of codes is fairly clear now (e.g. not found is a client error, similar to a 404).

That depends very much on the application. Not found is the canonical example of where this is fuzzy as it can be a system error due to a propagation delay or a user error due to requesting a non-existent resource.

> Neither of the other parameters is dynamic, and the labels are stable (service names and methods).

The question isn't stability, it's a question of cardinality. 10+ services/methods is not unusual.

> We both know from our previous SRE experience how important good RPC metrics are, as developers often forget to instrument their code in other fashion.

I know that good RPC metrics are important, and that average is sufficient for the vast majority of use cases.

I also know that histograms metrics can get massive, and in fact are the top cause of overloaded monitoring from my experience. In the system you're thinking of histograms were disabled by default (with a warning about metric size), and even then weren't broken out by return code.

> And not initialize the variables in func init of the package? Can can you give an example of how you guys do it?

I don't think there's any examples currently. It'd probably be a separate function call by the user that does the registration.

**Comment by mwitkow** [comment]
I'm still not convinced by the argument of prefixing `service` with `grpc_service`. I do believe that people who use gRPC think of their gRPC services as services, since that is what `.proto` files and gRPC generated code is reminding them about on every step. I think they'd be more willing to change their target names just to have their own internal naming consistency. Especially since the target names are usually configured by someone other than the developers that implement the `service` code and consume the metrics in Prometheus. I think we should cater to the developers that write the code itself, if we want Prometheus adoption to move into the mainstream.

I'd like to hear your thoughts on what `type` should be (with values `unary`, `client_stream`, `server_stream`, `bidi_stream`). I agree it's not great, but I really have no better candidate.  

@fabxc I've addressed the `code` problem in the histogram. The histograms no longer have them.
@brian-brazil The histograms are now enabled by the user explicitly, and `code` representation is retained in the `handled_total` CounterVec.

From my point of view, I think we've reached a good middle ground. @fabxc @brian-brazil, would you agree?

**Comment by brian-brazil** [comment]
> I think they'd be more willing to change their target names just to have their own internal naming consistency. Especially since the target names are usually configured by someone other than the developers that implement the service code and consume the metrics in Prometheus.

I'm not following your argument here. As you note the developers of a piece of instrumentation are unlikely to have control of every environment in which it's used. To me that means they won't generally be trivially able to change the target labels just because one small library happens to clash. In some organisations it may not even be a possibility, as the label could already could be too well established to change.

I've worked with such instrumentation in the past, it was very difficult to deal with. This is why the guidelines recommend avoiding any labels as instrumentation labels that are likely to be used as target labels.

> I think we should cater to the developers that write the code itself, if we want Prometheus adoption to move into the mainstream.

That's a nice idea, it doesn't work out in practice as there's far more users of instrumentation than the person who implemented it. All instrumentation is part of a wider context. If every developer makes  locally-optimal decisions, the overall ecosystem suffers.

For the users of this code to have the best experience and adoption, you need them all using the same labels so they can share expressions, dashboards and alerts. `service` is not an instrumentation label that'll allow your users to do this. `grpc_service` will.

> and code representation is retained in the handled_total CounterVec.

All the possible code labels still need to be initialised.

**Comment by mwitkow** [comment]
> > and code representation is retained in the handled_total CounterVec.
> All the possible code labels still need to be initialised.

As I said above: sadly it's not possible this is not possible from a technical point of view. These interceptors have no access to the gRPC server's registration (these are private) and can't enumerate each `service` and `method` in order to initialize things to 0.

> > I'm not following your argument here. As you note the developers of a piece of instrumentation are unlikely to have control of every environment in which it's used. To me that means they won't generally be trivially able to change the target labels just because one small library happens to clash. In some organisations it may not even be a possibility, as the label could already could be too well established to change.

> > I've worked with such instrumentation in the past, it was very difficult to deal with. This is why the guidelines recommend avoiding any labels as instrumentation labels that are likely to be used as target labels.

My argument here is that gRPC is not "one small library", it is meant as "the way your services interact". Given that the concept of `service` in gRPC is not moveable, I'm pretty certain that organizations will accommodate it in their lexicon if they choose to adopt gRPC. As such I think it is more likely case is that monitoring targets would cease to use `services` or at least introduce a [metric labelling](https://prometheus.io/docs/operating/configuration/#<metric_relabel_config>) for them.

`service` is nowhere mentioned in Prometheus documentation as a recommended target label (`job` [is](https://prometheus.io/docs/concepts/jobs_instances/)) nor in the documentation for [target labels to avoid](https://prometheus.io/docs/instrumenting/writing_exporters/#labels). Even for Kubernetes the label is  `service_name` for its targets.

I appreciate your vast experience, but on the matter of `service` vs `grpc_service` you seem to be offering opinions and not facts. At this point I'm going to keep it as `service` since making it `grpc_service` would entail moving other labels (`method`, `code`) to have the prefix (for consistency), and I think that would look out of place with many other things.

**Comment by brian-brazil** [comment]
> As I said above: sadly it's not possible this is not possible from a technical point of view.

They could be initialised on every RPC. If we can't do that then we need to create a success and failure metric without labels, which will avoid the main pitfalls (with the downside of loss of granularity).

> My argument here is that gRPC is not "one small library", it is meant as "the way your services interact". 

That's making the assumption that the user is aware that gRPC exists and has explicitly chosen to use it at an organisational level from day one. For example etcd uses it internally, and I'm willing to bet that a good chunk of users don't know that. There's discussions about Prometheus using it, and the majority of our users wouldn't know that it exists other than as a obscure implementation detail.

I don't think that it's reasonable to ask that just because a sysadmin pulls in a random binary that they need to make cascading changes across their entire monitoring system.

In the grand scheme of things, whatever it's ambitions, gRPC is one small library. There are many many other libraries and frameworks with the goal of being "the way your services interact".

> Given that the concept of service in gRPC is not moveable

I'm not arguing that the concept is moveable, I'm arguing that it needs to be namespaced to avoid other similarly named immovable concepts.

> or at least introduce a metric labelling for them.

Metric relabelling is only intended as a stopgap while you fix the instrumentation. In this case it'd still leave your users unable to share expressions, dashboards and alerts which is a bad user experience.

> service is nowhere mentioned in Prometheus documentation as a recommended target label (job is) 

Neither is team or env - but those are also unwise choices as instrumentation labels.

Job is special, it is effectively a required label.

> nor in the documentation for target labels to avoid

The list is not intended to be exhaustive.

> Even for Kubernetes the label is `service_name` for its targets.

The full name is `__meta_kubernetes_service_name`.

> I appreciate your vast experience, but on the matter of service vs grpc_service you seem to be offering opinions and not facts.

I'm basing this on direct experience from using Stubby, which had this exact problem with this exact label for some of its metrics. Googlers and Xooglers will use `service` as a target or external label out of habit (and likely many others too as it's an obvious target label), and this library using it as an instrumentation label will cause them problems.

I've also interacted with a number of other systems which had clashes on common target labels (`shard` being the usual one), and it made the metrics effectively unusable.

> grpc_service would entail moving other labels (method, code) to have the prefix (for consistency), and I think that would look out of place with many other things.

Yes, you'd also need to change method at least.

**Comment by ejona86** [comment]
> `type` seems wonky, but I still don't have a better candidate

Yeah... grpc-java is using `Method Type`. C++ is using `RPC Type`. Not happy with it but it was "good enough." `Streaming Type` may not be too wonky. `Cardinality` may not be too bad.

**Comment by brian-brazil** [comment]
> Yeah... grpc-java is using Method Type. C++ is using RPC Type. Not happy with it but it was "good enough." Streaming Type may not be too wonky. Cardinality may not be too bad.

Sometimes you just end up with `type` being the best option.

**Comment by mwitkow** [comment]
@ejona86 thanks for clarifying. It seems that `type` is the only common denominator across languages, so I'll roll with that.

@brian-brazil, we both know that Googlers/Xooglers would use `job` out of habit..  That's why Prometheus uses it :wink: 

I tried the metrics with `grpc_` prefix on our systems today. The expressions look wonky and repetitive, and I'd rather stick with `service` for now. The moment someone encounters this problem in a real world outside of our bike-shedding session here, I promise I'll add the option to for the labels to be `grpc_` prefixed. I hope you're ok with that.

I moved the `0` initialisation to a separate issue #5, as I'd like to have some more clarity.

Closing this for now.

**Comment by brian-brazil** [comment]
> @brian-brazil, we both know that Googlers/Xooglers would use job out of habit.. That's why Prometheus uses it ￼

`job` is one of the labels they must use anyway, as Prometheus maintains it as a special label along with `instance`.

`zone`, `service` and `shard` are the others which which will be used out of habit, however they are not special within Prometheus as we don't presume that everyone uses a particular label taxonomy.

> The moment someone encounters this problem in a real world outside of our bike-shedding session here

I'm surprised that me experiencing this exact problem in the past is not considered real world.

> I promise I'll add the option to for the labels to be grpc_ prefixed. I hope you're ok with that.

The core of my argument is that we need to avoid fragmentation of the labels within the userbase, this doesn't prevent that fragmentation.

**Comment by mwitkow** [comment]
> The core of my argument is that we need to avoid fragmentation of the labels within the userbase, this doesn't prevent that fragmentation.

@brian-brazil, if you're so worried about label clash, why doesn't Prometheus implement a label namespacing system (e.g. using first metric prefix for all labels)? Or blacklist certain labels in Prometheus client libraries? Or have clear documentation front and centre about which labels are prohibited from use with a clear explanation why? Surely the whole cohesion of the userbase can't rely on your free time to have a watchful eye on all the small libraries making sure they all are vigilant of this problem.

> I'm surprised that me experiencing this exact problem in the past is not considered real world.

My argument was (and still is) that users that rely on gRPC and other sevice-frameworks (as you rightfully pointed out) would not name their targets `service`, since a `service` would already have a meaning for them that doesn't imply a single running process that runs somewhere.

I'm sorry, but your argument about your prior experience isn't really concrete, and AFAIK can't relate to my argument above. This is the first gRPC Prometheus library out there. Both Prometheus and gRPC are relatively new technologies. The intersection of users of both is tiny. I think that users of both will not encounter a clash (exactly how Google services didn't by using `job` as the "service" name). The moment someone comes with a setup that would prove the above point wrong, I'll add the prefixes.

**Comment by brian-brazil** [comment]
> if you're so worried about label clash, why doesn't Prometheus implement a label namespacing system (e.g. using first metric prefix for all labels)?

If it were being designed today something similar to Kubernetes' labels is likely what we'd have. We don't have that though and it's too late to add it, so we get to emulate it by hand.

> Or blacklist certain labels in Prometheus client libraries?

We blacklist the internal reserved names, however there's use cases for using target labels in instrumentation (primarily around alerting) so we can't blacklist them.

> Or have clear documentation front and centre about which labels are prohibited from use with a clear explanation why?

There's a few which are more important to avoid, which are clearly documented.

> Surely the whole cohesion of the userbase can't rely on your free time to have a watchful eye on all the small libraries making sure they all are vigilant of this problem.

That's one of the big reasons I wrote the exporter guidelines. This is just one of many things that needs to be thought about when writing instrumentation.

> I'm sorry, but your argument about your prior experience isn't really concrete

I'm not sure how I can make it more concrete for you. There was a special organisational-level label called `service`, and this interacted very badly with a library that used instrumentation labels also called `service`. This was not the only time I came across such an issue.

> The intersection of users of both is tiny.

I don't think we should be doing engineering based on the premise that neither technology in question is going to grow it's userbase.

Even ignoring that, this library is likely to be used in etcd, which will be used by all Kubernetes users, who directly provide Prometheus instrumentation and are also associated with the CNCF. That means the intersection is going to be at least a good chunk of Kubernetes users.

> exactly how Google services didn't by using job as the "service" name

Your recollection is incomplete, see the SRE book page 112. `job` is not the service name, it is the job name. `service` is the service name.

> The moment someone comes with a setup that would prove the above point wrong, I'll add the prefixes.

Per above, Google is such a setup. I expect them to pick up this code via etcd, and thus there's likely to be a clash.

**Comment by mwitkow** [comment]
I'll add the `grpc_` prefix to all the labels. 
Primarily because @ejona86 mentioned that `grpc_type` is one of the names for "unary" etc., and I gave that a second thought.
Secondarily: indeed, my memory served me wrong.The Google internal name is `rpc_service_name`. Thirdly, I would like the library and its conventions (with a Java version coming soonish) to be used more widely among the Prometheus community. A disagreement with a maintainer, even if it boils down to "your opinion vs mine", wouldn't help here.

**Comment by ejona86** [comment]
> Primarily because @ejona86 mentioned that grpc_type is one of the names for "unary" etc., and I gave that a second thought.

Not that it would change your decision, but _RPC_ type was one of the names, not gRPC.

**Comment by brian-brazil** [comment]
> A disagreement with a maintainer, even if it boils down to "your opinion vs mine", wouldn't help here.

Indeed. We have client library and exporter guidelines for exactly this reason. In this case the code is probably sufficient with a statement to replicate it exactly.
